helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -413,6 +413,31 @@ public void addRoutingTableChangeListener(
     _routingTableChangeListenerMap.put(routingTableChangeListener, new ListenerContext(context));
     logger.info("Attach RoutingTableProviderChangeListener {}",
         routingTableChangeListener.getClass().getName());
+    if (_sourceDataTypeMap.isEmpty()) {
+      routingTableChangeListener.onRoutingTableChange(getRoutingTableSnapshot(), context);
+    } else {
+      for (PropertyType propertyType : _sourceDataTypeMap.keySet()) {
```

Comment:
As Junkai mentioned, I think we need a new interface for the listeners anyway.
The current one won't tell the provider which type the listener wants to listen to. So I think we just add a new API so as to add the listeners with the right type.
In this case, the old API needs to be deprecated and if user add their listen using this API, we shall just assign it to all the valid types.

Reply:
Why would we need the type as input? We already put the type in _sourceDataTypeMap when users initialize the routing table provider, right? So we can trigger callback with required type of snapshot.

Reply:
For example, this constructor allows the user to create a map of types.
```
  public RoutingTableProvider(HelixManager helixManager, Map<PropertyType, List<String>> sourceDataTypeMap) {
    this(helixManager, sourceDataTypeMap, true, DEFAULT_PERIODIC_REFRESH_INTERVAL);
  }
```
But we don't have a way for the user to specify the listener for different types. So one listener will be triggered multiple times if multiple types are specified.

Reply:
I see your point, so right now although the routing table provider only initialized with certain source data types, users will still get callback on all snapshot change, and they need to filter based on the type to see whether the change is they want. But I think this change is beyond the scope of this PR. It's always a problem after we have multiple types, all the Zookeeper change will trigger callback, if we want to change, we will need to distinguish between different listeners, and the change also needs to be made in `notifyRoutingTableChange` when we receive a change. 
For this PR only, we do trigger callback based on types, and user won't get unused callback for those types they don't care. For more general type support for callback, we'll need more changes. Thoughts?

Reply:
I agree this is out of this PR's scope. My main concern is that this PR adds one more place that we need to modify later. It would be great if we have only one method to trigger all onRoutingTableChange calls.

Reply:
Yeah, the onRoutingTableChange is called in both add listener and whenever there is any change triggers it. These two won't be able to be merged, right?